src/inet/routing/leach/Leach.h

```diff
@@ -0,0 +1,125 @@
+//
+// SPDX-License-Identifier: LGPL-3.0-or-later
+//
+
+#ifndef INET_ROUTING_LEACH_LEACH_H_
+#define INET_ROUTING_LEACH_LEACH_H_
+
+#include <string.h>
+#include <vector>
+
+#include "inet/common/INETDefs.h"
+#include "inet/common/packet/Packet.h"
+#include "inet/networklayer/contract/IInterfaceTable.h"
+#include "inet/networklayer/contract/ipv4/Ipv4Address.h"
+#include "inet/networklayer/ipv4/IIpv4RoutingTable.h"
+#include "inet/networklayer/ipv4/Ipv4Header_m.h"
+#include "inet/networklayer/ipv4/Ipv4InterfaceData.h"
+#include "inet/networklayer/ipv4/Ipv4RoutingTable.h"
+#include "inet/routing/base/RoutingProtocolBase.h"
+#include "inet/routing/leach/LeackPkts_m.h"
+#include "inet/mobility/contract/IMobility.h"
+#include "inet/common/geometry/common/Coord.h"
+#include "inet/mobility/base/StationaryMobilityBase.h"
+#include "inet/power/contract/ICcEnergyStorage.h"
+#include "inet/common/lifecycle/LifecycleController.h"
+
+using namespace std;
+
+namespace inet {
+namespace leach {
+
+class INET_API Leach: public RoutingProtocolBase {
+
+private:
+    NetworkInterface *wirelessInterface = nullptr;
+    unsigned int sequencenumber = 0;
+    cModule *host = nullptr;
+    Ipv4Address idealCH;
+    cMessage *event = nullptr;
+    int round = 0;
+    ModuleRefByPar<IInterfaceTable> interfaceTable;
+
+protected:
+    simtime_t dataPktSendDelay;
+    simtime_t CHPktSendDelay;
+    simtime_t roundDuration;
+
+    int numNodes = 0;
+    double clusterHeadPercentage = 0.0;
+
+    // object for storing CH data per node in vector array
+    struct nodeMemoryObject {
+        Ipv4Address nodeAddr;
+        Ipv4Address CHAddr;
+        double energy;
+    };
+
+    struct TDMAScheduleEntry {
+        Ipv4Address nodeAddress;
+        double TDMAdelay;
+    };
+
```

Comment:
Are you sure vector<> is the best data structure for this? As far as I see, you are often looking up using an ip address and iterate over the whole vector. This implies an o(2) complexity with node number. A hash-map would be probably a better approach (I'm not sure on this)

Reply:
I looked into this and it does seem that vectors
So something like `unordered_map<Ipv4Address, nodeMemoryObject> nodeMemory;` yes?

With a modified nodeMemory struct looking something like this to avoid having redundant nodeAddress value.
```
struct nodeMemoryObject {
        Ipv4Address CHAddr;
        double energy;
};
```

Reply:
Yes, but this is really just optimization so it can be done at the end of the project. Use whichever structure is more natural to you.

Reply:
Understood. I'll keep it as it is for now because it seems like I'll have to change several other components to accommodate this. So will keep this suggestion open to take care of in future.

I'm now moving on to integrating the UDP app with the LEACH module for data packet transmission.
Will update below.